### DISCUSSION TITLE
YourTTS Training Time

### DISCUSSION THREAD

**Discussion by Rico2000** [coqui-ai/TTS#1235] [General Q&A] [answered]
Hey,

I have a problem with the Yourtts model. I'm trying to train the model in German. For this I use the dataset Multilingual LibriSpeech http://www.openslr.org/94/. 
The problem is the training time. One epoch took about 5 hours with a batch size of 8. I have trained on 4 Navidia V100 and I suspect that some settings are not correct. 

`> Setting up Audio Processor...
 | > sample_rate:16000
 | > resample:True
 | > num_mels:80
 | > log_func:np.log
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:0
 | > mel_fmax:None
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:False
 | > trim_db:23.0
 | > do_sound_norm:False
 | > do_amp_to_db_linear:False
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:2.718281828459045
 | > hop_length:256
 | > win_length:1024
 | > Found 469942 

 > initialization of speaker-embedding layers.
 > Using CUDA:  True
 > Number of GPUs:  4
 > `speakers_file` is updated in the config.json.

 > Model has 85480893 parameters
[4m[1m > EPOCH: 0/1000[0m
 > DataLoader initialization
 | > Use phonemes: False
 | > Number of instances : 469942
 | > Max length sequence: 219246.5
 | > Min length sequence: 35174.5
 | > Avg length sequence: 132966.10140400304
 | > Num. instances discarded by max-min (max=160000, min=32768) seq limits: 89103
 | > Batch group size: 0.`

Logs: [std_log.txt](https://github.com/coqui-ai/TTS/files/8111752/std_log.txt)
GPU Utilization: ![GPU_Utilization](https://user-images.githubusercontent.com/60230024/155017952-d4f4d1ee-5dd0-4d5b-92a6-7101b53ded70.JPG)
GPU Memory usage: ![GPU_Memory_usage](https://user-images.githubusercontent.com/60230024/155018078-c0af94e8-572d-4ec9-9fd7-fea9cd862760.JPG)

Does anybody know what is wrong and how can i speed up training?

Thanks Best Regards
Rico

**Comment by erogol** [comment, marked as answer]
with what command do you start the training?

**Reply by Rico2000** [reply]
that is the code

`dataset_config = [
    BaseDatasetConfig(name = "test",meta_file_train="train",meta_file_val="dev",  path=data_folder)
]

audio_config = BaseAudioConfig(
    sample_rate=16000,
    win_length=1024,
    hop_length=256,
    num_mels=80,
    preemphasis=0.0,
    ref_level_db=20,
    log_func="np.log",
    do_trim_silence=False,
    trim_db=23.0,
    mel_fmin=0,
    mel_fmax=None,
    spec_gain=1.0,
    signal_norm=True,
    do_amp_to_db_linear=False,
    resample=True,
)

vitsArgs = VitsArgs(
    use_language_embedding=False,
    use_speaker_embedding=True,
    use_sdp=False,
)

config = VitsConfig(
    model_args=vitsArgs,
    audio=audio_config,
    run_name="vits_mailabs",
    use_speaker_embedding=True,
    batch_size=64,
    eval_batch_size=32,
    batch_group_size=0,
    num_loader_workers=4,
    num_eval_loader_workers=4,
    run_eval=True,
    test_delay_epochs=-1,
    epochs=1000,
    text_cleaner="basic_german_cleaners",
    use_phonemes=False,
    phoneme_language="de-de",
    phoneme_cache_path=os.path.join(output_path, "phoneme_cache"),
    compute_input_seq_cache=True,
    print_step=25,
    use_language_weighted_sampler=False,
    print_eval=False,
    mixed_precision=False,
    sort_by_audio_len=True,
    min_seq_len=32 * 256 * 4,
    max_seq_len=160000,
    output_path=output_path,
    datasets=dataset_config,
    characters={
        "pad": "_",
        "eos": "&",
        "bos": "*",
        "characters": "!¡'(),-.:;¿?abcdefghijklmnopqrstuvwxyzµßöäü «°±µ»$%&‘’‚“`”„",
        "punctuations": "!¡'(),-.:;¿? ",
        "phonemes": None,
        "unique": True,
    }
)
ap = AudioProcessor(**config.audio.to_dict())

train_samples, eval_samples = load_tts_samples(dataset_config, eval_split=True, formatter = format_ml)
print(len(train_samples))
print(len(eval_samples))


speaker_manager = SpeakerManager()
speaker_manager.set_speaker_ids_from_data(train_samples + eval_samples)
config.model_args.num_speakers = speaker_manager.num_speakers

model = Vits(config, speaker_manager)

trainer = Trainer(
    args,
    config,
    output_path,
    model=model,
    train_samples=train_samples,
    eval_samples=eval_samples,
    training_assets={"audio_processor": ap},
     parse_command_line_args = False
)
trainer.fit()`

an then in command line python TTS/bin/distribute.py --script train.py

**Reply by erogol** [reply]
You need to use CUDA_VISIBLE_DEVICES

https://tts.readthedocs.io/en/latest/training_a_model.html

**Reply by Rico2000** [reply]
i know  i run a seperate config so enable all gpu's. While training all GPU's are over 50% usage. The Problem appear idenpendet from the number of GPU's. I think Training shoulf be faster on GPU than like 5 hours per epoch or am i wrong? The dataset hast about 390.000 samples. Is there any configuartion which can lead to faster Training?
One thing is that i can only run on batch_size 16, otherwise Out of Memory happens. In there paper they mentioned a batch_size 64 with a V100.

**Reply by erogol** [reply]
What is the step time printed on the terminal?

**Reply by Rico2000** [reply]
Hey,
steptime was around 0.8 - 1.2. I appended logs and screenshots to my original question above.
Hope this helps

Thanks

**Reply by erogol** [reply]
0.8 - 1.2 is expected. Then your training speed is fine.

**Reply by Rico2000** [reply]
oke thanks. Last Question whats about the batch_size? I oly can use a batch_size of 16 with a sequenc length of 10 sec and sample rate of 10. Same happens if i only use one V100
